### PR TITLE
Implement single connection benchmark

### PR DESCRIPTION
**Pull request by sylwiaszunejko**
API for hooking to `Write()` and `Read()` of `Conn` struct added. New struct `TrafficRecorder` added for enabling recording and mocking single Connection flow. 

Fixes: #296

### CODE REVIEW

**Comment by sylwiaszunejko** [comment]
Adding this as a draft, because I wanted to make sure this part looks like intended in the issue, before pushing the rest.

**Comment by sylwiaszunejko** [comment]
Pushed some changes, it is clearly work in progress with logging etc. But I am not sure why the test is not working and if this is a good approach. Some guidance will be appreciated

**Comment by sylwiaszunejko** [comment]
@dkropachev could you take a look?

File: conn.go [hunk 1]
```diff
@@ -166,9 +166,9 @@ var TimeoutLimit int64 = 0
 // queries, but users are usually advised to use a more reliable, higher
 // level API.
 type Conn struct {
-	conn net.Conn
-	r    *bufio.Reader
-	w    contextWriter
+	connWrapper *ConnWrapper
+	r           *bufio.Reader
+	w           contextWriter
```
Comment (dkropachev) [inline conn.go:171]: Let's start with proper abstraction.
Let's have a one interface, call it `NetworkConnection`:
```
type NetworkConnection interface {
	SetWriteDeadline(time.Time) error
	WriteContext(ctx context.Context, p []byte) (n int, err error)
	io.Reader
	io.Closer
	RemoteAddr() net.Addr
}
```

It is going to be used as following:
```
type Conn struct {
	nc NetworkConnection
}
```

Let's have factory on the `ClusterConfig`
```
type NetworkConnectionFactory interface {
	Wrap(conn net.Conn) NetworkConnection
}
```

Now, `DefaultNetworkConnectionFactory` would be wrapping writer with `deadlineContextWriter` and reader with `bufio.NewReader`

**Review by dkropachev — CHANGES_REQUESTED** [review]

**Comment by sylwiaszunejko** [comment]
@dkropachev I pushed some changes, it is clearly wip, but I want to make sure the direction is right

File: tests/bench/bench_conn_test.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+		err = session.Query(`SELECT pk, ck, v FROM single_conn_bench.table1 WHERE pk = ?;`, i).WithContext(ctx).Consistency(gocql.One).Scan(&pk, &ck, &v)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func BenchmarkSingleConnection(b *testing.B) {
+	// RecordTraffic(100)
```
[the remark this answers is not in the document]
Comment (sylwiaszunejko) [reply tests/bench/bench_conn_test.go:69]: `NewSingleHostQueryExecutor` should be used for recording traffic or replaying it?

[the remark this answers is not in the document]
Comment (dkropachev) [reply tests/bench/bench_conn_test.go:69]: For both

File: scylla.go [hunk 3]
```diff
@@ -795,6 +798,116 @@ func (d *ScyllaShardAwareDialer) DialContext(ctx context.Context, network, addr
...
+	if writeErr != nil {
+		return n, fmt.Errorf("failed to record read: %w", writeErr)
+	}
+	_, writeErr = c.fd_reads.Write([]byte("\n"))
+	if writeErr != nil {
+		return n, fmt.Errorf("failed to record read: %w", writeErr)
+	}
+
+	return n, err
+}
```
[the remark this answers is not in the document]
Comment (sylwiaszunejko) [reply scylla.go]: pushed some changes to the previous comments, but still working on properly keeping track of request/repsonse relation

File: dialer/recorder/recorder.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	return n, err
+}
+
+func (c *ConnectionRecorder) Write(b []byte) (n int, err error) {
+	n, err = c.orig.Write(b)
+
+	// Only record if EXECUTE frame
+	if b[4] != 0x0A {
```
Comment (dkropachev) [inline dialer/recorder/recorder.go:135]: We should record every frame.

File: dialer/recorder/recorder.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+		return n, err
+	}
+
+	if c.write_record.new {
+		stream_id := int(b[2])<<8 | int(b[3])
+		b[2] = byte('0')
+		b[3] = byte('0')
+		writeHash := murmur.Murmur3H1(b[:9])
```
Comment (dkropachev) [inline dialer/recorder/recorder.go:143]: I believe hash should be calculated in a different way for different frame types.
It is better to make a function of it `getFrameHash(frame []byte) int64` to be used here and on replay

File: dialer/recorder/recorder.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+func (c *ConnectionRecorder) Write(b []byte) (n int, err error) {
+	n, err = c.orig.Write(b)
+
+	// Only record if EXECUTE frame
+	if b[4] != 0x0A {
+		return n, err
+	}
+
+	if c.write_record.new {
```
Comment (dkropachev) [inline dialer/recorder/recorder.go:139]: Frames are read sequentially, but written in parallel to reads, you should have a list of all the submitted frame here.

File: dialer/recorder/recorder.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+		b[2] = byte('0')
+		b[3] = byte('0')
+		writeHash := murmur.Murmur3H1(b[:9])
+
+		c.write_record.record = make(map[string]interface{})
+		c.write_record.record["stream_id"] = stream_id
+		c.write_record.record["hash"] = writeHash
+		c.write_record.record["data"] = b[:n]
+
+		c.write_record.to_record = int(b[8])
```
Comment (dkropachev) [inline dialer/recorder/recorder.go:150]: Frame lengh is 4 bytes

**Review by dkropachev — CHANGES_REQUESTED** [review]
Probably you could reuse something from `TestServer` from `conn_test.go` to replay traffic

File: dialer/recorder/recorder.go (hunk 6 above)
Comment (sylwiaszunejko) [reply dialer/recorder/recorder.go:141]: That's why I have reparate records for reads and writes, I am not sure why would we need a list in recorder

File: dialer/recorder/recorder.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+	if c.write_record.new {
+		p := 4
+		stream_id := int(b[2])
+		if b[0] > 0x02 {
+			p = 5
+			stream_id = int(b[2])<<8 | int(b[3])
+		}
+
+		c.write_record.record = make(map[string]interface{})
+		c.write_record.to_record = int(b[p+0])<<24 | int(b[p+1])<<16 | int(b[p+2])<<8 | int(b[p+3])
```
Comment (dkropachev) [inline dialer/recorder/recorder.go:150]: lenght is length of body without header, so you need to add `p + 3` to it

File: dialer/utils.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+			stringLenght := int(frame[index])<<8 | int(frame[index+1])
+			index = index + 2 + stringLenght
+			bytesLength := int(frame[index+0])<<24 | int(frame[index+1])<<16 | int(frame[index+2])<<8 | int(frame[index+3])
+			index = index + 4 + bytesLength
+		}
+		fmt.Println("INDEX: ", index)
+		statementLength := int(frame[index+0])<<24 | int(frame[index+1])<<16 | int(frame[index+2])<<8 | int(frame[index+3])
+		fmt.Println("STMT LEN: ", statementLength)
+		fmt.Println("FRAME LEN: ", len(frame))
+		return murmur.Murmur3H1(frame[index : statementLength+4])
```
Comment (dkropachev) [inline dialer/utils.go:65]: You need to exclude timestamp, if it is there (`if flags & flagDefaultTimestamp > 0` ). Take a look at `framer.writeQueryParams` for more details: https://github.com/scylladb/gocql/blob/57a9143688f0e4932333a32a4b61534a974a6364/frame.go#L1570-L1583

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: dialer/utils.go (hunk 9 above)
Comment (sylwiaszunejko) [reply dialer/utils.go:65]: for opQuery I only take statement in consideration, I omit whole QueryParams part so there is no need to exclude timestamp, should I take queryParams into the hash?

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: dialer/replayer/replayer.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+	stream_ids_to_replay []int
+	index                int
+	partial              int
+}
+
+func (c *ConnectionReplayer) Read(b []byte) (n int, err error) {
+	if c.index < 0 || c.index >= len(c.query_ids_to_replay) {
+		return 0, nil
+	}
+	response := c.queries[c.query_ids_to_replay[c.index]].Response[c.partial:]
```
[the remark this answers is not in the document]
Comment (sylwiaszunejko) [reply dialer/replayer/replayer.go:68]: sometimes there is a Read before a Write is executed and we don't want to fail in such case just return 0 and wait until something arrives

File: dialer/utils.go (hunk 9 above)
Comment (dkropachev) [reply dialer/utils.go]: Yes, I believe you should

File: dialer/replayer/replayer.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (dkropachev) [reply dialer/replayer/replayer.go]: Realy ? It should not happen, if it is happening it means that there is some problem with implementation.

File: dialer/replayer/replayer.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+		fmt.Println(err)
+		return nil, err
+	}
+
+	conn, err = dialerWithLocalAddr.DialContext(ctx, network, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewConnectionReplayer(path.Join(d.dir, fmt.Sprintf("%s-%d", addr, sourcePort)), conn)
```
Comment (dkropachev) [inline dialer/replayer/replayer.go:45]:
```suggestion
	return NewConnectionReplayer(path.Join(d.dir, fmt.Sprintf("%s-%d", addr, sourcePort)))
```

File: dialer/replayer/replayer.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (dkropachev) [reply dialer/replayer/replayer.go]: Ok, I give it a test, it indeed does that, sure, let's leave it as it is

**Comment by dkropachev** [comment]
@sylwiaszunejko , we need to disable these guys for `NewSingleHostQueryExecutor`
```
		s.hostSource = &ringDescriber{session: s}
		s.ringRefresher = newRefreshDebouncer(ringRefreshDebounceTime, func() error { return refreshRing(s.hostSource) })
```

File: dialer/replayer/replayer.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+func (c *ConnectionReplayer) Write(b []byte) (n int, err error) {
+	writeHash := dialer.GetFrameHash(b)
+	for i, q := range c.frames {
+		if q.Hash == writeHash {
+			c.pushStreamIDToReplay(b, i)
+			return len(b), nil
+		}
+	}
+	return 0, fmt.Errorf("unable to find a response to replay")
```
Comment (dkropachev) [inline dialer/replayer/replayer.go:113]: I see that there is a problem here, driver will silently handle this error by reconnecting to the host,so, it is better to panic here to make sure we have all the response that we need.

**Review by dkropachev — DISMISSED** [review]

File: dialer/replayer/replayer.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	} else {
+		b[2] = byte(stream)
+	}
+}
+
+func (c *ConnectionReplayer) Read(b []byte) (n int, err error) {
+	frame := c.getPendingFrame()
+	if frame == nil {
+		return 0, nil
+	}
```
[the remark this answers is not in the document]
Comment (sylwiaszunejko) [reply dialer/replayer/replayer.go:86]: @dkropachev  gotRequest should be `chan struct{}` in `ConnectionReplayer`? what should `c.closed()` do?

[the remark this answers is not in the document]
Comment (dkropachev) [reply dialer/replayer/replayer.go:86]: > @dkropachev gotRequest should be `chan struct{}` in `ConnectionReplayer`? 

True 
> what should `c.closed()` do?

It should return true if connection is closed

[the remark this answers is not in the document]
Comment (sylwiaszunejko) [reply dialer/replayer/replayer.go]: I am not sure how to properly do it, it keeps getting stuck, I pushed the way I tried use this channel  (it is commented)

File: dialer/replayer/replayer.go [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+	frame := c.getPendingFrame()
+	for frame == nil {
+		// fmt.Println("FRAME NIL")
+		// <-c.gotRequest
+		// if c.Closed() {
+		// 	return 0, io.EOF
+		// }
+		// frame = c.getPendingFrame()
+		return 0, nil
+	}
```
Comment (dkropachev) [inline dialer/replayer/replayer.go:94]:
```suggestion
	for frame == nil {
		<-c.gotRequest
		frame = c.getPendingFrame()
	}
	if c.Closed() {
		return 0, io.EOF
	}
```

File: exec.go [hunk 15]
```diff
@@ -51,7 +51,6 @@ func NewSingleHostQueryExecutor(cfg *ClusterConfig) (e SingleHostQueryExecutor,
 
 	// If protocol version not set assume 4 and skip discovery
 	if c.ProtoVersion == 0 {
-		e.session.logger.Print("gocql: setting protocol version 4")
```
Comment (dkropachev) [inline exec.go:54]: Why ?

File: dialer/replayer/replayer.go [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+	c.frameIdx = c.frameIdx + 1
+	c.frameResponsePosition = 0
+	return len(response), err
+}
+
+func (c *ConnectionReplayer) Write(b []byte) (n int, err error) {
+	writeHash := dialer.GetFrameHash(b)
+	// defer func() {
```
Comment (dkropachev) [inline dialer/replayer/replayer.go:115]: It is better to move this code to the end of `pushStreamIDToReplay`, and please make it non-blocking

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: dialer/recorder/recorder.go [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+	net.Dialer
+}
+
+func (d *RecordDialer) DialContext(ctx context.Context, network, addr string) (conn net.Conn, err error) {
+	fmt.Println("Dial Context Record Dialer")
+	sourcePort := gocql.ScyllaGetSourcePort(ctx)
+	fmt.Println("Source port: ", sourcePort)
+	// if sourcePort == 0 {
+	// 	return d.Dialer.DialContext(ctx, network, addr)
+	// }
```
Comment (dkropachev) [inline dialer/recorder/recorder.go:34]: a left over ?

File: tests/bench/bench_single_conn_test.go [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+			})
+		}
+	})
+}
+
+func init() {
+	InitializeCluster()
+	RecordInsertTraffic(10, "/home/sylwiaszunejko/gocql/tests/bench/rec_insert")
+	RecordSelectTraffic(10, "/home/sylwiaszunejko/gocql/tests/bench/rec_select")
+}
```
Comment (dkropachev) [inline tests/bench/bench_single_conn_test.go:156]: Please check if it is possible to make it run only if you do `go test --update-golden`.

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: exec.go (hunk 15 above)
Comment (sylwiaszunejko) [reply exec.go:54]: e.session is not created yet, it is nil so this cause panic it is initialized here: https://github.com/scylladb/gocql/blob/master/exec.go#L67

File: tests/bench/bench_single_conn_test.go (hunk 18 above)
Comment (sylwiaszunejko) [reply tests/bench/bench_single_conn_test.go:156]: I tried to add a flag like in `recreate_test.go`, but it complains that the flag is not defined, pls remind me how to provide that flag in `recreate_test.go` to make it work? I tried `/integration.sh integration scylla --update-golden` but it does not work

Comment (dkropachev) [reply tests/bench/bench_single_conn_test.go:156]: Try this:
```
func TestMain(m *testing.M) {
	update := flag.Bool("update-golden", false, "Update golden files")
	flag.Parse()
	if *update {
		InitializeCluster()
		RecordInsertTraffic(10, "rec_insert")
		RecordSelectTraffic(10, "rec_select")
	}
	os.Exit(m.Run())
}

```

File: exec.go (hunk 15 above)
Comment (dkropachev) [reply exec.go:54]: Ohh, crazy, so it was broken all this time.

**Review by dkropachev — DISMISSED** [review]
Final change I it is finally ready, great job, it produces pretty stable results:
```
cpu: 12th Gen Intel(R) Core(TM) i9-12900HK
BenchmarkSingleConnectionInsert
BenchmarkSingleConnectionInsert/Insert
BenchmarkSingleConnectionInsert/Insert/Case0
BenchmarkSingleConnectionInsert/Insert/Case0-20           242354              4332 ns/op
BenchmarkSingleConnectionInsert/Insert/Case1
BenchmarkSingleConnectionInsert/Insert/Case1-20           295485              4334 ns/op
BenchmarkSingleConnectionInsert/Insert/Case2
BenchmarkSingleConnectionInsert/Insert/Case2-20           304800              4332 ns/op
BenchmarkSingleConnectionInsert/Insert/Case3
BenchmarkSingleConnectionInsert/Insert/Case3-20           266346              4240 ns/op
BenchmarkSingleConnectionInsert/Insert/Case4
BenchmarkSingleConnectionInsert/Insert/Case4-20           254448              4042 ns/op
BenchmarkSingleConnectionInsert/Insert/Case5
BenchmarkSingleConnectionInsert/Insert/Case5-20           288975              4270 ns/op
BenchmarkSingleConnectionInsert/Insert/Case6
BenchmarkSingleConnectionInsert/Insert/Case6-20           279169              4152 ns/op
BenchmarkSingleConnectionInsert/Insert/Case7
BenchmarkSingleConnectionInsert/Insert/Case7-20           294770              4074 ns/op
BenchmarkSingleConnectionInsert/Insert/Case8
BenchmarkSingleConnectionInsert/Insert/Case8-20           299186              3994 ns/op
BenchmarkSingleConnectionInsert/Insert/Case9
BenchmarkSingleConnectionInsert/Insert/Case9-20           305224              4147 ns/op
PASS

```

File: tests/bench/bench_single_conn_test.go (hunk 18 above)
Comment (sylwiaszunejko) [reply tests/bench/bench_single_conn_test.go]: it works with this version

**Comment by dkropachev** [comment]
@sylwiaszunejko , I spot another problem, standard benchrmark library does not collect `goroutine` stats, which means that some parts are not shown in results.
Unfortunately due to the library limitations, there is only one way to fix it is to come up with custom test library, which is a bit cumbersome.
Given all that I think it is better to merge it and look for solution that would show everything later, WDYT?

**Review by dkropachev — APPROVED** [review]

**Comment by sylwiaszunejko** [comment]
> @sylwiaszunejko , I spot another problem, standard benchrmark library does not collect `goroutine` stats, which means that some parts are not shown in results. Unfortunately due to the library limitations, there is only one way to fix it is to come up with custom test library, which is a bit cumbersome. Given all that I think it is better to merge it and look for solution that would show everything later, WDYT?

I agree

**Comment by sylwiaszunejko** [comment]
I created an issue to collect goroutine stats https://github.com/scylladb/gocql/issues/330

**Pull request merged**